### DISCUSSION TITLE
Is My Approach to Inserting Rows in Supabase Secure and Correct in Terms of Relationships, auth and RLS?

### DISCUSSION THREAD

**Discussion by vladimirzb** [supabase/supabase#16888] [Questions] [answered]
I am working on a Next.js 13 project using Supabase for the backend. I have implemented a feature to add "ingresos" (income entries) to a table. I would like to know if my approach is secure, especially in terms of how I am inserting the user ID, and if I am handling the relationships correctly.

### Code:

Here's the client component code:

```jsx
'use client';
import { useState, useEffect } from 'react';
import { addIngreso } from '../actions/add-ingreso';
import { createClientComponentClient } from '@supabase/auth-helpers-nextjs';
import { Database } from '@/types/supabase';

const AddIngreso = () => {
  const supabase = createClientComponentClient<Database>();
  const [userId, setUserId] = useState<string | null>(null);
  const [amount, setAmount] = useState('');
  const [source, setSource] = useState('');

  useEffect(() => {
    const fetchUser = async () => {
      const { data } = await supabase.auth.getUser();
      setUserId(data?.user?.id ?? null);
    };

    fetchUser();
  }, []);

  const handleAddIngreso = async () => {
    if (userId === null) {
      console.error("User ID is null. Cannot add ingreso.");
      return;
    }

    const timestamp = new Date().toISOString();

    const { data, error } = await addIngreso(Number(amount), source, userId, timestamp);

    if (error) {
      console.error("Error adding ingreso:", error);
      return;
    }

    console.log("Ingreso added:", data);
  };

  return (
    <div>
      <input 
        type="number" 
        placeholder="Amount" 
        value={amount} 
        onChange={(e) => setAmount(e.target.value)} 
      />
      <input 
        type="text" 
        placeholder="Source" 
        value={source} 
        onChange={(e) => setSource(e.target.value)} 
      />
      <button onClick={handleAddIngreso}>Add Ingreso</button>
    </div>
  );
};

export default AddIngreso;
```

And the server action code:

```jsx
'use server';

import { createServerActionClient } from "@supabase/auth-helpers-nextjs";
import { cookies } from 'next/headers'
import { Database } from '@/types/supabase'

export async function addIngreso(amount: number, source: string, userId: string, timestamp: string) {
  const supabase = createServerActionClient<Database>({ cookies });
  const { data, error } = await supabase
    .from('ingresos')
    .insert([
      { 
        cantidad: amount, 
        fuente: source,
        user_id: userId,
        timestamp_with_time_zone: timestamp
      },
    ])
    .select();

  if (error) {
    return { error };
  }

  return { data };
}
```

### RLS and Relationships:

I have set up an RLS policy as follows:

```sql
(auth.uid() = user_id)
```
![image](https://github.com/supabase/supabase/assets/55926702/45ab27f4-843a-4abe-9169-c3a4205ee472)

I also have a foreign key relationship between `ingresos.user_id` and `auth.users.id`.

### Security Concern:

My biggest concern is whether the way I am inserting the user ID is secure. 

### Relationship Concern:

I would also like to know if I am handling the relationships correctly during the insert operation. Is my foreign key setup and the way I am inserting the `user_id` in line with best practices?

### Additional Security Clarification:

From what I understand, the RLS policy I've set up should already provide a layer of security. Specifically, even if someone were to get their hands on a user ID, they shouldn't be able to insert a row unless the `auth.uid()` matches the `user_id` in the `ingresos` table, as per my RLS policy:

```sql
(auth.uid() = user_id)
```

However, I would like to be absolutely sure that this is the case. Can someone confirm if my understanding of RLS in this context is correct and if it provides the security layer I assume it does?

**Comment by GaryAustin1** [comment, marked as answer]
If you have RLS `auth.uid() = user_id` for insert/update policy then it must be the signed in user's id in that column data.
But you can also make the default auth.uid() for that column then you don't need to pass it in.  You still need the policy though to protect as default can be replaced by data provided in update/insert.

**Reply by vladimirzb** [reply]
Thank you for your help!
